pgjdbc/src/main/java/org/postgresql/hostchooser/GlobalHostStatusTracker.java

```diff
@@ -62,12 +62,14 @@ private static boolean updateStatusFromTo(HostStatus oldStatus, HostStatus newSt
     synchronized (hostStatusMap) {
       for (HostSpec hostSpec : hostSpecs) {
         HostSpecStatus hostInfo = hostStatusMap.get(hostSpec);
-        // return null status wrapper if if the current value is not known or is too old
-        if (hostInfo == null || hostInfo.lastUpdated < latestAllowedUpdate) {
+        // return null status wrapper if if the current value is not known
+        if (hostInfo == null) {
           hostInfo = new HostSpecStatus(hostSpec, null, Long.MAX_VALUE);
         }
         // candidates are nodes we do not know about and the nodes with correct type
-        if (hostInfo.status == null || targetServerType.allowConnectingTo(hostInfo.status)) {
+        if (hostInfo.status == null
+            || hostInfo.lastUpdated < latestAllowedUpdate
```

Comment:
The thing is due to "host recheck", it should forget known master/slave status of the host. So it ends up with (host1==master, host2==slave, host3==unknown, host4==unknown) kind of states.
In that case, for loadBalance=true; connect=preferSlave  it should try connecting to unknown hosts and it should try its best finding a slave there (between unknowns and known-slaves).

As far as I can see, the change was "never reset status", and I do not see why it is a good change.

Reply:
This change just move "hostInfo.lastUpdated < latestAllowedUpdate" from line 66 to line 70, when "loadBalance=true&targetServerType=preferSlave", the host whose status is expired or unknown do have the chance to be chosen.
In other words, the behavior is same as before when "loadBalance=true".

But when "loadBalance=false", this change made the hosts with the right matched status(whether expired or not) have higher priority than others to be chosen.
That can avoid some useless connecting retry.

Reply:
> the host whose status is expired or unknown do have the chance to be chosen.

They do not unless the requested status matches server state.
For instance:
1) requested is "master"
2) once the server was identified as "slave"
3) then pgjdbc would never try to connect to that server since it remembers the status as slave.

Am I missing something?

Reply:
>1. requested is "master"
>2. once the server was identified as "slave"
>3. then pgjdbc would never try to connect to that server since it remembers the status as slave.

That's true only when the server's status is not expired(default is 10s)

when status was expired, the server will be add to the candidates list.

GlobalHostStatusTracker.java
```
if (hostInfo.status == null
   || hostInfo.lastUpdated < latestAllowedUpdate
   || targetServerType.allowConnectingTo(hostInfo.status)) {
       candidates.add(hostInfo);
    }
```

and then have chance to be chosen.

MultiHostChooser.java
```
    if (!loadBalance) {
      sortCandidates(candidates);
    } else {
      shuffleCandidates(candidates);
    }
```